### PR TITLE
[757] Support can add new school users

### PR DESCRIPTION
**Pull request by asmega**
### Context

- https://trello.com/c/T2A6JdPn/757-add-a-new-school-user-via-support

### Changes proposed in this pull request

- New page in support to add new school users
- Tweaks to `CreateUserService` to remove overbearing overrides

### Guidance to review

- Login as support user
- Navigate to a school via RB
- Should be a button below users table to add new users
- Click this button
- Should be taken to page to add new user to the school
- Fill it in incorrectly
- Should see errors
- Fix mistakes and submit
- Should create this new user on the school
- Should send email to new user

### CODE REVIEW

File: app/services/create_user_service.rb [hunk 1]
```diff
@@ -1,23 +1,17 @@
 class CreateUserService
   def self.invite_responsible_body_user(user_params = {})
-    user = User.new(user_params.merge(override_params))
+    user = User.new(user_params)
```
Comment (benilovj) [inline app/services/create_user_service.rb:3]: I don't think we don't want to be making this change to this method. "All responsible body users get a TechSource account" is a part of the domain logic that correctly lived in the service layer. There isn't any use case for expositing either `ordered_devices` or `approved_at` outside the service.

File: app/services/create_user_service.rb [hunk 2]
```diff
@@ -1,23 +1,17 @@
...
     if user.save
       InviteResponsibleBodyUserMailer.with(user: user).invite_user_email.deliver_later
     end
     user
   end
 
   def self.invite_school_user(user_params = {})
-    user = User.new(user_params.merge(override_params))
+    user = User.new(user_params)
```
Comment (benilovj) [inline app/services/create_user_service.rb:11]: I think you've found a bug here... it looks like previously, even if a new school user was invited without a TechSource account, one was still being created for them. I'd suggest we fix the bug as part of this PR.

So for `self.invite_school_user` I think we want to let outside callers to set `orders_devices` but not `approved_at`.

File: app/views/support/devices/schools/show.html.erb [hunk 3]
```diff
@@ -50,6 +50,8 @@
   <p class="govuk-body">None</p>
 <% end %>
 
+<%= link_to 'Add school user', new_support_devices_school_user_path(school_urn: @school.urn), class: 'govuk-button' %>
```
Comment (benilovj) [inline app/views/support/devices/schools/show.html.erb:53]:
```suggestion
<%= govuk_button_link_to 'Add school user', new_support_devices_school_user_path(school_urn: @school.urn) %>
```

File: config/locales/en.yml [hunk 4]
```diff
@@ -100,6 +100,7 @@
       can_order_devices: Yes, give them access to the TechSource website
       can_order_devices_hint: This is where they’ll place orders.
       cannot_order_devices: 'No'
+    new_school_user: New school user
```
Comment (benilovj) [inline config/locales/en.yml:103]:
```suggestion
    new_school_user: Invite a new user
```
This copy is consistent with the responsible body and school versions of this apge.

**Review by benilovj — CHANGES_REQUESTED** [review]

File: config/locales/en.yml (hunk 4 above)
Comment (asmega) [reply config/locales/en.yml]: done

File: app/views/support/devices/schools/show.html.erb (hunk 3 above)
Comment (asmega) [reply app/views/support/devices/schools/show.html.erb]: good spot, i've changed this now

File: app/services/create_user_service.rb (hunk 1 above)
Comment (asmega) [reply app/services/create_user_service.rb]: I see, I've changed `invite_responsible_body_user ` back to the original

File: app/services/create_user_service.rb (hunk 2 above)
Comment (asmega) [reply app/services/create_user_service.rb]: Made a tweak, should now work as you've described

File: app/controllers/support/devices/users_controller.rb [hunk 5]
```diff
@@ -1,4 +1,21 @@
...
+
+  def create
+    @school = School.find_by(urn: params[:school_urn])
+    user_attributes = @school.users.build(user_params).attributes
+    @user = CreateUserService.invite_school_user(user_attributes)
+
+    if @user.persisted?
+      redirect_to support_devices_school_path(urn: @school.urn)
+    else
+      render :new
```
Comment (benilovj) [inline app/controllers/support/devices/users_controller.rb:15]:
```suggestion
      render :new, status: :unprocessable_entity
```

File: app/views/support/devices/users/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+<%- title = t('page_titles.new_school_user') %>
+<%- content_for :title, title_with_error_prefix(title, @user.errors.any?) %>
+<%- content_for :before_content do %>
+  <%= breadcrumbs([
+    { t('page_titles.support_responsible_bodies') => support_devices_responsible_bodies_path },
+    { @school.responsible_body.name => support_devices_responsible_body_path(@school.responsible_body) },
+    { @school.name => support_devices_school_path(urn: @school.urn) },
+    'New school user',
```
Comment (benilovj) [inline app/views/support/devices/users/new.html.erb:8]:
```suggestion
    title,
```

File: app/views/support/devices/users/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <h1 class="govuk-heading-xl">
+      <%= title %>
+    </h1>
+
+    <%= form_for @user, url: support_devices_school_users_path(school_urn: @school.urn) do |f| %>
+      <%= f.govuk_error_summary %>
+      <%= render partial: 'shared/school_user_form', locals: { form: f } %>
+      <%= f.govuk_submit 'Create school user' %>
```
Comment (benilovj) [inline app/views/support/devices/users/new.html.erb:21]:
```suggestion
      <%= f.govuk_submit 'Send invite' %>
```

**Review by benilovj — APPROVED** [review]
Looks fine now, just a couple of minor suggestions for design consistency

File: app/controllers/support/devices/users_controller.rb (hunk 5 above)
Comment (asmega) [reply app/controllers/support/devices/users_controller.rb]: done

File: app/views/support/devices/users/new.html.erb (hunk 7 above)
Comment (asmega) [reply app/views/support/devices/users/new.html.erb]: done

File: app/views/support/devices/users/new.html.erb (hunk 6 above)
Comment (asmega) [reply app/views/support/devices/users/new.html.erb]: done

**Comment by benilovj** [comment]
Just one more thing. We want the add user button to be above the Users list, like so:

image

**Comment by asmega** [comment]
> Just one more thing. We want the add user button to be above the Users list, like so:
> 
> image

Outside this ticket I think there are other places that will also need changing if we want this as the pattern eg. on the RB show page

**Pull request merged**